### PR TITLE
[dynamo][moco] Disallow_in_graph distributed APIs

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100199
* __->__ #100071

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by anijain2305** [comment]
@wconstab Wondering if you know of a better way here, instead of just graph breaking on them?

**Review by jansel — APPROVED** [review]

**Comment by yanboliang** [comment]
> The missing piece is how to ensure we graph-break on the full set of them. I thought I stamped a PR (maybe from @yanboliang) a while back that skipped a large chunk of torch.distributed- looks like that did not land or else I can't find remnants of it.

Yes, that PR has been reverted multiple times, I finally use regex matching to skip ```torchrec.distribued```(make it support ```torch.package```), but only be used internally behind ```is_fbcode```. I think we should skip these files right now, not only because they are collectives ops, but also they have memory leak issue due to their implementation.

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -124,6 +124,11 @@ def _export_c_types():
 global _c10d_error_logger
 _c10d_error_logger = _get_or_create_logger()
 
+# There is no clear definition of torch.distributed ops. This helper set allows
+# TorchDynamo to selectively disallow all the distributed ops from the Fx
+# graphs.
+distributed_c10d_ops = set()
```
Comment (anijain2305) [inline torch/distributed/distributed_c10d.py:130]: @wconstab Is this too hacky? Problem is there is no easy way to tell what's a torch "op" and whats not.

Comment (awgu) [reply torch/distributed/distributed_c10d.py:130]: Maybe cc @H-Huang @kwen2501 -- do we expect a function to use the `exception_handler` decorator if and only if it is a c10d op?

Comment (wconstab) [reply torch/distributed/distributed_c10d.py:130]: hah yea its kinda hacky.

But otoh i think it's reasonable to have a set of `dynamo_unsupported_distributed_c10d_ops` and explicitly add them all.  We should just get folks like @H-Huang to agree on the set of ops to initially tag.

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -4115,3 +4116,35 @@ def _get_group_tag(pg: ProcessGroup) -> str:
     if tag.startswith("user:"):
         tag = tag[5:]
     return tag
+
+
+# This ops are not friently to TorchDynamo. So, we decide to disallow these ops
+# in FX graph, allowing them to run them on eager, with torch.compile.
+dynamo_unsupported_distributed_c10d_ops = [
```
Comment (H-Huang) [inline torch/distributed/distributed_c10d.py:4123]: I am missing the context here. Is the idea just to exclude all the operations defined in `distributed_c10d.py`?

Comment (anijain2305) [reply torch/distributed/distributed_c10d.py:4123]: Yes

Comment (anijain2305) [reply torch/distributed/distributed_c10d.py:4123]: Or specifically, the operators that do not work well with TorchDynamo tracing and we want to graph break (i.e. fallback to eager) on them

Comment (anijain2305) [reply torch/distributed/distributed_c10d.py:4123]: @H-Huang ping

Comment (H-Huang) [reply torch/distributed/distributed_c10d.py:4123]: I see, then would it be possible to just get all operations of the module through something like:
https://stackoverflow.com/questions/139180/how-to-list-all-functions-in-a-module

Then we don't have to create this list in `distributed_c10d` and worry about keeping it up to date

Comment (anijain2305) [reply torch/distributed/distributed_c10d.py:4123]: Yes, so we tried that but then it gives a large number of functions and many of them not really "operators".

Another way I tried earlier in a commit - https://github.com/pytorch/pytorch/pull/100071/commits/8e6dee36ff3eea0648222ca739458ebf140cb4a1 
was to record the ops in exception_handler. The assumption was that anyone adding a new op will decorate it withe exception handler. But, that looked little hacky as well.

Comment (anijain2305) [reply torch/distributed/distributed_c10d.py:4123]: Thanks for pointing that out. Removed it from the list.

**Review by H-Huang — APPROVED** [review]
LGTM

**Pull request closed without merging**